lib/net/imap.rb

```diff
@@ -1530,6 +1530,9 @@ def start_tls_session(params = {})
       end
       @sock = SSLSocket.new(@sock, context)
       @sock.sync_close = true
+      unless hostname.nil?
+        @sock.hostname = hostname
+      end
```

Comment:
For what it's worth, `@host` is already being used in this method (just a few lines down) without being passed in as an arg. I think it's safe to assume that it's set. So my patch (which I hadn't gotten around to submitting, so thanks!) simply adds the following line:

```diff
         context.verify_callback = VerifyCallbackProc
       end
       @sock = SSLSocket.new(@sock, context)
+      @sock.hostname = @host if context.verify_mode != VERIFY_NONE
       @sock.sync_close = true
       @sock.connect
       if context.verify_mode != VERIFY_NONE
```

Reply:
Although I probably should not have made it conditional on verify_mode, but should have instead checked to be sure host is not nil, as you did.

Reply:
I notice that net/http _always_ sets the address, conditional only on whether `hostname=` exists (which is I suppose for backwards compatibility with pre-SNI versions of OpenSSL). Probably best to follow that approach.
https://github.com/ruby/ruby/blob/9065336/lib/net/http.rb#L991